### PR TITLE
Build returns a Result

### PR DESCRIPTION
**Pull request by Robo210**
- LayerBuilder::build() now returns a Result rather than panics
  - Add new error type
  - Update samples
- Remove global_filter feature
  - Global filter is now a runtime option: LayerBuilder::build_global_filter

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum EtwError {
+    #[error("Provider group GUID must not be zeroes")]
```
Comment (idigdoug) [inline src/error.rs]: "zeroes" is a verb. (Internet says it is also a less-common spelling for the plural of the noun zero.)
"zeros" is noun. (Internet says it is the more-common spelling.)

File: src/error.rs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum EtwError {
+    #[error("Provider group GUID must not be zeroes")]
+    EmptyProviderGroupGuid,
+    #[error("Provider group names must be lower case ASCII or numeric digits: {0:?}")]
+    InvalidProviderGroupCharacters(String),
+    #[error("Linux provider names must be ASCII alphanumeric: {0:?}")]
+    InvalidProviderNameCharacters(String)
```
Comment (idigdoug) [inline src/error.rs]: My code is somewhat permissive about provider names. I recommend that the user stick to ASCII alphanumeric + underscores, but I only error-out for:

- Longer than 234 characters
- Contains NUL, space, or colon.

The user_events provider name validation is here:
https://github.com/microsoft/LinuxTracepoints-Rust/blob/3e4caad232e8671852e01e8a50a9cb5e9739d330/eventheader_macros/src/provider_info.rs#L48C9-L65C10

As well as ensuring that `provider_name.len() + group_name.len() < 234`.

File: src/layer.rs [hunk 3]
```diff
@@ -128,24 +126,23 @@ where
...
-        match &self.provider_group {
-            None => (),
-            Some(value) => Mode::assert_valid(value)
-        }
-
+    fn validate_config(&self) -> Result<(), EtwError> {
         #[cfg(target_os = "linux")]
         if self
             .provider_name
             .contains(|f: char| !f.is_ascii_alphanumeric())
```
Comment (idigdoug) [inline src/layer.rs]: Should definitely allow underscores.

**Review by idigdoug — DISMISSED** [review]

File: src/layer.rs [hunk 4]
```diff
@@ -128,13 +128,24 @@ where
...
+                .contains(|f: char| !f.is_ascii_alphanumeric() && f != '_')
+            {
+                // The perf command is very particular about the provider names it accepts.
+                // The Linux kernel itself cares less, and other event consumers should also presumably not need this check.
+                return Err(EtwError::InvalidProviderNameCharacters(self.provider_name.clone()));
+            }
+
+            let group_name_len = match &self.provider_group {
+                None => 0,
+                Some(ref name) => Mode::get_provider_group(&name).as_ref().len()
```
Comment (Robo210) [inline src/layer.rs]: I'm OK with saying len() is the number of characters, since we require it to be ASCII characters elsewhere.

**Comment by Robo210** [comment]
Global_filter feature will need to return. The current change always provides an implementation of Layer::enabled() and Layer::event_enabled() that check the current provider enablement status for that event. This will cause the events to always be enabled/disabled globally, which is undesirable. The correct way to do this is to use the default implementation of these methods when you intend on coexisting with other layers, which is what the feature flag controlled.

**Comment by idigdoug** [comment]
Without digging in, I don't immediately see the best way to do this, but I would still like this to be a runtime choice if at all possible, even if it means having two types of Layer, one with global filter and one without.

File: src/layer.rs (hunk 4 above)
Comment (idigdoug) [reply src/layer.rs]: str.len() returns the number of bytes, which is perfectly acceptable here.

https://doc.rust-lang.org/std/primitive.str.html#method.len

**Comment by Robo210** [comment]
> Without digging in, I don't immediately see the best way to do this, but I would still like this to be a runtime choice if at all possible, even if it means having two types of Layer, one with global filter and one without.

I think the Layer trait should be split by the tracing team into "trait for functions that compose with other layers" and "trait for functions that override things globally" (register_callsite, enabled, event_enabled,...). Without this, the best option is to just use features to make sure the consumer knows they are going to enable something that has global impact on every layer and is almost certainly not expected behavior.

**Review by idigdoug — APPROVED** [review]

**Pull request merged**